File: python-algorithm/ds_5_tree/tree_3_bst_property/L501.py
# -*- coding: utf-8 -*-
# @Time    : 2020/10/18 15:56
# @Author  : No Name

# 优化了空间复杂度，在遍历的过程中确定，不是将全部元素记录到数组中，再处理数组
class Solution:
    def findMode(self, root: TreeNode) -> List[int]:
        res = []
        if not root:
            return res
        pre = 0
        idx, cnt, max_freq = 0, 0, 0
        stk = []  # 模拟栈
        while stk != [] or root != None:
            while root:
                stk.append(root)
                root = root.left
            root = stk.pop(-1)
            idx += 1

            if idx == 1:
                pre = root.val
                cnt = 1
                max_freq = 1
            else:
                if root.val != pre:
                    if cnt > max_freq:
                        res = [pre]
                    elif cnt == max_freq:
                        res.append(pre)
                    max_freq = max(max_freq, cnt)
                    pre = root.val
                    cnt = 1
                else:
                    cnt += 1
            root = root.right

        # 最后的元素未作比较
        if cnt > max_freq:
            return [pre]
        elif cnt == max_freq:
            res.append(pre)
        return res

File: python-algorithm/ds_5_tree/tree_3_bst_property/L501.cpp

#include <iostream>
#include <vector>
using namespace std ;

struct  TreeNode {
    int val ;
    TreeNode *left ;
    TreeNode *right ;
    TreeNode(int x) : val(x) , left(NULL) , right(NULL) {}
};

// 题目要求不使用额外的空间
// (1)递归栈不能算吧 （2）用来装返回值的数组不能算
// 严格来讲下面空间复杂度，中间可能有开辟额外空间，不算O(1)
// 一定要O(1)空间复杂度的话，就做2次遍历，第一次记载下最大出现次数，第二次按照最大次数存储众数

class Solution {
public:
    int tmp_cnt = 1;
    int final_cnt = 1;
    TreeNode* pre = NULL;
    vector<int> findMode(TreeNode* root) {
        vector<int> res;
        InorderTree(root, res);
        return res;
    }

    void InorderTree(TreeNode* root, vector<int>& nums){
        if(root == NULL) return;
        InorderTree(root->left, nums);
        if(pre){
            if(root->val == pre->val)
                ++tmp_cnt;
            else
                tmp_cnt = 1;
        }
        if (tmp_cnt > final_cnt){
            final_cnt = tmp_cnt;
            nums.clear();
            nums.push_back(root->val);
        }
        else if(tmp_cnt == final_cnt)
            nums.push_back(root->val);
        pre = root;
        InorderTree(root->right, nums);
    }
};